### PR TITLE
Fix binding of video frames to video placeholder in `InternVL` model

### PR DESCRIPTION
**Pull request by daskol**
This PR fixes broken frame patch indexing in `InternVLProcessor`. If batch is non-trivial (more than one sample) and contains videos and video placeholder in text then the processor returns `pixel_values` of wrong shape. Specifically, it returns `(total_num_frames - 1, ...)` instead of `(total_num_frames, ...)`. The simple code snippet to reproduces follows.

```python
proc = AutoProcessor.from_pretrained('OpenGVLab/InternVL3_5-1B-HF')
texts = [
    '<video>\nAre there any cyan objects that enter the scene?\nno',
    '<video>\nAre there any red spheres that enter the scene?\nno',
]
frames = np.ones((4, 448, 448, 3), np.float32)
inputs = proc(text=texts,
              videos=[frames, frames], return_tensors='pt',
              videos_kwargs={'size': (448, 448)})
print(inputs.pixel_values.shape)  # (7, 3, 448, 448)
assert inputs.pixel_values.shape[0] == 2 * frames.shape[0]  # FAIL: 7 != 8
```

The original lines are authored by @yonigozlan. Could you take a look?

FYI @Rocketknight1 @zucchini-nlp

### CODE REVIEW

File: src/transformers/models/internvl/processing_internvl.py [hunk 1]
```diff
@@ -206,31 +206,50 @@ def __call__(
...
-            video_inputs = self.video_processor(videos=videos, **output_kwargs["videos_kwargs"])
+            # Here, `videos` are batch of frames (images) that is a list of 4D
+            # arrays. Since spatial dimensions of a video in a batch are
+            # arbitrary, `videos` is a list. However, `video_pixel_values` is
+            # scaled/cropped/resized/etc in order to constitute a 5D array of
+            # frames.
+            #
+            # NOTE There is no multi-patch support per frame.
+            video_kwargs = {**output_kwargs["videos_kwargs"], 'return_tensors': 'pt'}
+            video_inputs = self.video_processor(videos=videos, **video_kwargs)
```
Comment (yonigozlan) [inline src/transformers/models/internvl/processing_internvl.py]: `return_tensors` should already be included in output_kwargs["videos_kwargs"]
```suggestion
            video_inputs = self.video_processor(videos=videos, **output_kwargs["videos_kwargs"])
```

File: src/transformers/models/internvl/processing_internvl.py [hunk 2]
```diff
@@ -206,31 +206,50 @@ def __call__(
...
+        video_num_patches_indices = np.array([0])
         if videos is not None:
-            video_inputs = self.video_processor(videos=videos, **output_kwargs["videos_kwargs"])
+            # Here, `videos` are batch of frames (images) that is a list of 4D
+            # arrays. Since spatial dimensions of a video in a batch are
+            # arbitrary, `videos` is a list. However, `video_pixel_values` is
+            # scaled/cropped/resized/etc in order to constitute a 5D array of
+            # frames.
+            #
+            # NOTE There is no multi-patch support per frame.
```
Comment (yonigozlan) [inline src/transformers/models/internvl/processing_internvl.py]:
```suggestion
```

File: src/transformers/models/internvl/processing_internvl.py [hunk 3]
```diff
@@ -206,31 +206,50 @@ def __call__(
...
-            # Obtain per frame information first and then flatten to (BS * T, ...)
-            num_frames_per_video = [len(video) for video in video_pixel_values]
-            video_num_patches = [1 for frames in num_frames_per_video for _ in range(frames)]
-            video_patch_indices = np.cumsum(num_frames_per_video)
-            video_num_patches_indices = np.cumsum(video_num_patches)
+            # Now, we need to build a multi-index on all patches in a batch
+            # across videos and frames. We need this data structure to
+            # efficiently build an input data with respect to prompt. The issue
+            # is that a prompt in a batch does not necessary refer to a
+            # corresponding video.
```
Comment (yonigozlan) [inline src/transformers/models/internvl/processing_internvl.py]: Too verbose and unclear

File: tests/models/internvl/test_processing_internvl.py [hunk 4]
```diff
@@ -377,3 +379,22 @@ def _test_apply_chat_template(
...
+        videos = [frames, frames]
+
+        processor = self.get_processor()
+        inputs = processor(text=texts[:batch_size], return_tensors='pt',
+                           videos=videos[:batch_size],
+                           videos_kwargs={'size': (448, 448)})
+
+        actual_num_frames = inputs.pixel_values.pixel_values.shape[0]
+        expected_num_frames = sum(x.shape[0] for x in videos[:batch_size])
+        assert actual_num_frames == expected_num_frames
```
Comment (yonigozlan) [inline tests/models/internvl/test_processing_internvl.py:408]: Thanks for adding a test! But let's find a better/more descriptive name for it

**Review by yonigozlan — APPROVED** [review]
Thanks for fixing! I suggested some small changes, then LGTM!

File: src/transformers/models/internvl/processing_internvl.py (hunk 1 above)
Comment (daskol) [reply src/transformers/models/internvl/processing_internvl.py]: Well, but it is `None` if it is not set explicitly to `'pt'`. I just checked now.

File: src/transformers/models/internvl/processing_internvl.py (hunk 2 above)
Comment (daskol) [reply src/transformers/models/internvl/processing_internvl.py]: Fixed.

File: src/transformers/models/internvl/processing_internvl.py (hunk 3 above)
Comment (daskol) [reply src/transformers/models/internvl/processing_internvl.py]: Fixed.

(I see. Only maintainers are allowed to leave a comment J)

File: tests/models/internvl/test_processing_internvl.py (hunk 4 above)
Comment (daskol) [reply tests/models/internvl/test_processing_internvl.py:408]: Fixed.

File: src/transformers/models/internvl/processing_internvl.py (hunk 1 above)
Comment (yonigozlan) [reply src/transformers/models/internvl/processing_internvl.py]: It might be because of the way you call the processor in your example:
```python
inputs = proc(text=texts,
              videos=[frames, frames], return_tensors='pt',
              videos_kwargs={'size': (448, 448)})
```
Here you can just use size and it will automatically be used in video_kwargs, with return_tensors="pt":
```python
inputs = proc(text=texts,
              videos=[frames, frames], 
              size=(448, 448),
              return_tensors='pt',
```
Or you can add the return_tensors arg to the videos_kwargs as well, otherwise it's overwritten to None, and not set to the global return_tensors:
```python
inputs = proc(text=texts,
              videos=[frames, frames], return_tensors='pt',
              videos_kwargs={'size': (448, 448), return_tensors='pt',})
```

**Comment by yonigozlan** [comment]
@daskol  Looks like some InternVL processor tests are failing, you can run them locally using:
`RUN_SLOW=1 python -m pytest tests/models/internvl/test_processing_internvl.py` ;)

File: src/transformers/models/internvl/processing_internvl.py (hunk 1 above)
Comment (daskol) [reply src/transformers/models/internvl/processing_internvl.py]: No, it tried different combinations for `return_tensors`. The original code fails if `return_tensors` is not specified on flattening `videos_pixel_values.faltten(0, 1)` in the end of this branch (i.e. the code implies that `return_tensors='pt'` is enforced).

You can easily reproduce this case with the code in the initial post.

File: src/transformers/models/internvl/processing_internvl.py [hunk 5]
```diff
@@ -206,31 +206,38 @@ def __call__(
...
+            video_kwargs = {**output_kwargs["videos_kwargs"], "return_tensors": "pt"}
+            video_inputs = self.video_processor(videos=videos, **video_kwargs)
             video_pixel_values = video_inputs.pop("pixel_values_videos")
 
-            # Obtain per frame information first and then flatten to (BS * T, ...)
-            num_frames_per_video = [len(video) for video in video_pixel_values]
-            video_num_patches = [1 for frames in num_frames_per_video for _ in range(frames)]
-            video_patch_indices = np.cumsum(num_frames_per_video)
-            video_num_patches_indices = np.cumsum(video_num_patches)
+            batch_size, num_frames, *_ = video_pixel_values.shape
```
Comment (yonigozlan) [inline src/transformers/models/internvl/processing_internvl.py]: Actually, I'm not sure this is a good idea, as there can be videos with different number of frames. In this case, `"return_tensors": "pt"`, will fail. let's keep the previous:
```python
num_frames_per_video = [len(video) for video in video_pixel_values]
```
And let's not force  `"return_tensors": "pt"`

Comment (daskol) [reply src/transformers/models/internvl/processing_internvl.py]: The original code has already forced (see `video_pixes_values.flatten(0, 1)` and comment above https://github.com/huggingface/transformers/pull/41237#discussion_r2392808678).

Comment (yonigozlan) [reply src/transformers/models/internvl/processing_internvl.py]: Yes my bad, but this is a problem I think.
In my opinion, it would be better to force `return_tensors` to `None` in the `video_processor` call, so that we can support batch processing of videos with different number of frames.
We could have something like this in the loop:
```python
            return_tensors = output_kwargs["videos_kwargs"].pop(return_tensors, None)
            video_inputs = self.video_processor(videos=videos, **output_kwargs["videos_kwargs"], return_tensors=None)
            video_pixel_values = video_inputs.pop("pixel_values_videos")
            num_frames_per_video = [len(video) for video in video_pixel_values]
            num_frames = sum(num_frames_per_video)  # total
            video_patch_indices = np.empty(batch_size + 1, int)
            video_patch_indices[0] = 0
            video_patch_indices[1:] = np.cumsum(num_frames_per_video)
            video_num_patches = [1] * num_frames
            video_num_patches_indices = np.empty(num_frames + 1, int)
            video_num_patches_indices[0] = 0
            video_num_patches_indices[1:] = np.cumsum(video_num_patches)
            # flatten
            video_pixel_values = [video_frame for video in video_pixel_values for video_frame in video]
            video_pixel_values = torch.stack(video_pixel_values) if return_tensors=="pt" else video_pixel_values 
```

What do you think @zucchini-nlp ? Am I missing something?

Comment (daskol) [reply src/transformers/models/internvl/processing_internvl.py]: I suggest to keep it as simple as possible and work on the better solution later.

From my perspective, the best solution is to consider content processing as a binding prepared statement to actual variable. I mean that we should to parse prompt, extract content placeholders and load content (e.g. video from a file) on demand. At the moment, all content is loaded greedily. It is quite common pattern in content processing. Hence, it should live in base processor.

Moreover, there is no proper support for multi-patch frames. This makes the situation even more complicated.

**Comment by daskol** [comment]
@yonigozlan Thanks for the oneliner to run tests. All jobs are green now.

File: src/transformers/models/internvl/processing_internvl.py (hunk 5 above)
Comment (zucchini-nlp) [reply src/transformers/models/internvl/processing_internvl.py]: I agree that we need to support variable number of frames when possible. Some models support it by patchifying or by padding

I am fine with leaving it since it has been working with torch-assumption for a while and not all models support variable frame count. Let's try to fix it in the subsequent PR though

File: src/transformers/models/internvl/processing_internvl.py [hunk 6]
```diff
@@ -206,31 +206,38 @@ def __call__(
...
             image_pixel_values = image_inputs.pop("pixel_values")
             image_num_patches_indices = np.cumsum(image_num_patches)
+
+        video_num_patches = []  # per frame
+        video_pixel_values = None
+        video_patch_indices = np.array([0])
+        video_num_patches_indices = np.array([0])
         if videos is not None:
-            video_inputs = self.video_processor(videos=videos, **output_kwargs["videos_kwargs"])
+            video_kwargs = {**output_kwargs["videos_kwargs"], "return_tensors": "pt"}
```
Comment (zucchini-nlp) [inline src/transformers/models/internvl/processing_internvl.py]: i think we should enforce it in the `ModelKwargs` defaults

Comment (daskol) [reply src/transformers/models/internvl/processing_internvl.py]: I didn't get it. What is `ModelKwargs`?

Comment (zucchini-nlp) [reply src/transformers/models/internvl/processing_internvl.py]: sorry, I meant the a typed dict - 'InternVLKwargs', which has defaults for each modality

Comment (daskol) [reply src/transformers/models/internvl/processing_internvl.py]: Fixed.

**Review by zucchini-nlp — APPROVED** [review]
Oke, lets merge it. Thanks

**Comment by daskol** [comment]
Thank you for your time.
